cwltool fixes

andreyto opened the pull request:
Fixes for a few problems revealed while running production workflows with `cwltool` on HPC cluster (Torque, NFS). Please see full commit messages.

cgcloud-jenkins commented:
Can one of the admins verify this patch?

cgcloud-jenkins commented:
Can one of the admins verify this patch?

mr-c commented:
Jenkins, okay to test

benedictpaten commented:
jenkins, test

benedictpaten commented:
jenkins, test this please

benedictpaten commented:
I do not love the use of the sleep in the filestore "exists" function - do you have a non-posix file system or are you covering up a race condition elsewhere in the code?

diekhans commented on src/toil/jobStores/fileJobStore.py:
```diff
@@ -121,8 +122,18 @@ def batch(self):
+                return False
+            logger.warning("Job file `{}` for job `{}` does not exist, retry #{} in {}s.".\
+                    format(jobFile,jobStoreID,iTry+1,sleepTime))
+            time.sleep(sleepTime)
```
ah, I see this is on  NFS.  This could be an issue with NFS caching.  I can be a real pain.  I might make sense to make the JobStore NFS-aware.

andreyto replied:
Totally agree that spin-waiting can be hiding bugs in the code. It did help with persistent tracebacks when running on the NFS, though. Can you tell me: is it possible in Toil design that the leader is checking for the existence of a job file that is supposed to be created on a different node (by a worker)? It is hard to figure it out by looking at the code. If the answer is Yes, and the shared FS is essentially used for synchronization, then this strongly supports the need for spin-waiting on the NFS. On our cluster, it usually takes about 15 sec for the files to appear in `ls` across nodes. IT played with various settings on the NetApp and nodes trying to make this delay shorter, and to no avail. Various other workflow systems have settings to control how long they should wait for the outputs after completion of a task.

hannes-ucsc replied:
IIRC, FileJobStore assumes Posix semantics. If the underlying file-system is only eventually consistent–and it appears that NFS is–then the need arises for another strongly consistent mechanism to tell the job store that a file will, definitely, but eventually appear. This is how the AWS job store works, it uses the strongly consistent SimpleDB for tracking which files exist. Merely, spin-locking would not work as it would introduce the possibility of dead lock: Toil relies on both the absence and the presence of files to trigger a progression in the workflow, I think.

A hacky workaround would be, as @andreyto suggests, to always wait for a configurable time for a file to appear or to disappear (!) but this would slow down the workflow. Maybe the disappearance delay can be limited to only those cases in which workflow progress is gated on the absence of a file.

diekhans replied:
The spin-wait seems a necessary evil to support NFS without a lot of investment.  NFS can be run in a coherent manner with at a significant and often unacceptable slowdown.

My concerns about this are
  - are there any other places where this kind of spin-wait is necessary?
  - do we need a config variable for the time?
  - the rationale is not documented in the code for those, like me, who have forgotten the pain of NFS.

ejacox replied:
I like @diekhans points. A config variable and better documentation (rationale) should be added.

For the other point, my understanding is that the fileJobStore was built to test workflows locally. There could well be other consistency issues. Ideally, these would be handled in an abstraction layer since this problem applies to the other jobStores as @hannes-ucsc pointed out.

Does anyone object to patching it until we have time for a better solution.

andreyto replied:
@hannes-ucsc , could you describe and/or point to places in the code where `Toil relies on <snip> the absence `?
As far as NFS consistency guarantees, they are described in the `man nfs` help page, as well as discussed in [this thread](https://serverfault.com/questions/638601/nfs-caching-issue).
In short, without config changes that kill the performance, the NFS has:
- no cache coherency between clients for open file writes/reads.  MPI applications with concurrent I/O to the same file are affected by this, and that leads to the development of parallel file systems like GPFS, Lustre, PVFS. Toil is not affected by this because the client/server pipe on the NFS always does full sync when the file is closed, and Toil does not do concurrent writes (I hope).
- no cache coherency for metadata operations (directory modifications). Toil is affected by this, but not all is bad. NFS (see the man page) provides strict time guarantees for how long it will take for the cache to be reconciled on the client. These can be tuned in `/etc/fstab`. Incidentally, the 30 sec max wait that I set in the spin-wait code is the default in the NFS man page. Most sites never change the default.

A number of workflow systems implement the spin-wait for this very pattern (Makeflow, Galaxy, Cromwell), and do just fine. Basically, the system learns that the job has finished when its disappears from `qstat` output, and the system waits for the output files to appear on the NFS. Consider also that the batch job scheduling system has its own delays in starting jobs, so the extra worst-case 30 sec (with default NFS setup) of wait will be incremental to the queue delays (probably double the delay).

It is possible of course to implement some other notification mechanism like running a 0MQ server process in the leader, but is it really worth the trouble?

I agree that the spin-wait time should be made a tunable parameter.

hannes-ucsc replied:
> could you describe and/or point to places in the code where

https://github.com/BD2KGenomics/toil/blob/27edc4b64e992f29de64c863f00044771d494022/src/toil/leader.py#L834

is a place where the deletion of a file is used as a signal. Or

https://github.com/BD2KGenomics/toil/blob/27edc4b64e992f29de64c863f00044771d494022/src/toil/leader.py#L732

is an important condition where the absence of a file moves the workflow forward.

And this is just the result of a cursory scan of the code base. I think every job store file access would have to be scrutinized for the effects of eventual consistency.

> NFS (see the man page) provides strict time guarantees 

Whenever I hear the words "NFS", "strict" and "guarantee" in the same sentence, I smile. But maybe NFS has gotten better. I haven't used it in a while.

> A number of workflow systems implement the spin-wait for this very pattern (Makeflow, Galaxy, Cromwell), 

Do they also use the file system to manage the workflow state? As I mentioned before: if you know through an out-of-band channel that a file *should* (*should not*) exist, then its trivial to just wait until it appears (disappears). But Toil tracks the workflow itself in the job store by persisting the very job objects to it, not just the inputs and outputs of those jobs.

> running a 0MQ server process in the leader, but is it really worth the trouble

Toil workflows are resumable from nothing more but the information in the job store, so relying on ephemeral messages to communicate information about files out of band may be difficult.

I want to caution against fixing one particular workflow to deal with NFS's eventual consistency and concluding that that fixes every other workflow on every other installation of NFS.

Maybe the Toil team should experiment with an eventually consistent file system, or at least emulate one and run the job store unit tests on it. The difficulty lies in the fact that NFS's eventual consistency can only be observed with more than one node.

hannes-ucsc replied:
By the way, I don't speak for the Toil team. I'm just a former contributor. The above is merely my own opinion.

cgcloud-jenkins commented:
I worry about this somewhat - if we go down this route of patching the
filestore comprehensively we'll need to do this for essentially all the
filestore methods to cope with eventual consistency. That said, if this one
change fixes things almost all the time, and we provide some comments to
explain the rationale then I'd be 1+. The right way to do this is probably
with a synchronization layer, which would probablt mean a database. I very,
very much dislike that option because it brings many issues with
scalability and management.

On Tue, Feb 20, 2018 at 9:25 AM, Edwin Jacox <notifications@github.com>
wrote:

> *@ejacox* commented on this pull request.
> ------------------------------
>
> In src/toil/jobStores/fileJobStore.py
> <https://github.com/BD2KGenomics/toil/pull/2068#discussion_r169395690>:
>
> > @@ -121,8 +122,18 @@ def batch(self):
>              self.update(jobGraph)
>          self._batchedJobGraphs = None
>
> -    def exists(self, jobStoreID):
> -        return os.path.exists(self._getJobFileName(jobStoreID))
> +    def exists(self, jobStoreID,maxTries=1):
> +        sleepTime = 1
> +        for iTry in range(1,maxTries+1):
> +            jobFile = self._getJobFileName(jobStoreID)
> +            if os.path.exists(jobFile):
> +                return True
> +            if iTry >= maxTries:
> +                return False
> +            logger.warning("Job file `{}` for job `{}` does not exist, retry #{} in {}s.".\
> +                    format(jobFile,jobStoreID,iTry+1,sleepTime))
> +            time.sleep(sleepTime)
>
> I like @diekhans <https://github.com/diekhans> points. A config variable
> and better documentation (rationale) should be added.
>
> For the other point, my understanding is that the fileJobStore was built
> to test workflows locally. There could well be other consistency issues.
> Ideally, these would be handled in an abstraction layer since this problem
> applies to the other jobStores as @hannes-ucsc
> <https://github.com/hannes-ucsc> pointed out.
>
> Does anyone object to patching it until we have time for a better solution.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/2068#discussion_r169395690>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMXKjmcJJWF4QAhdCcT6XNM9b0dzXoGxks5tWv_1gaJpZM4SHi3T>
> .
>
> --
> You received this message because you are subscribed to the Google Groups
> "cgloud-jenkins" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to cgloud-jenkins+unsubscribe@googlegroups.com.
> To post to this group, send email to cgloud-jenkins@googlegroups.com.
> To view this discussion on the web visit https://groups.google.com/d/
> msgid/cgloud-jenkins/BD2KGenomics/toil/pull/2068/
> review/97915785%40github.com
> <https://groups.google.com/d/msgid/cgloud-jenkins/BD2KGenomics/toil/pull/2068/review/97915785%40github.com?utm_medium=email&utm_source=footer>
> .
>
> For more options, visit https://groups.google.com/d/optout.
>



-- 
Benedict
(calendar invites: bpaten@ucsc.edu,
 appointments: Tina Bernard <tibernar@ucsc.edu>)

benedictpaten commented:
On Tue, Feb 20, 2018 at 7:37 PM, Andrey Tovchigrechko <
notifications@github.com> wrote:

> *@andreyto* commented on this pull request.
> ------------------------------
>
> In src/toil/jobStores/fileJobStore.py
> <https://github.com/BD2KGenomics/toil/pull/2068#discussion_r169528200>:
>
> > @@ -121,8 +122,18 @@ def batch(self):
>              self.update(jobGraph)
>          self._batchedJobGraphs = None
>
> -    def exists(self, jobStoreID):
> -        return os.path.exists(self._getJobFileName(jobStoreID))
> +    def exists(self, jobStoreID,maxTries=1):
> +        sleepTime = 1
> +        for iTry in range(1,maxTries+1):
> +            jobFile = self._getJobFileName(jobStoreID)
> +            if os.path.exists(jobFile):
> +                return True
> +            if iTry >= maxTries:
> +                return False
> +            logger.warning("Job file `{}` for job `{}` does not exist, retry #{} in {}s.".\
> +                    format(jobFile,jobStoreID,iTry+1,sleepTime))
> +            time.sleep(sleepTime)
>
> @hannes-ucsc <https://github.com/hannes-ucsc> , could you describe and/or
> point to places in the code where Toil relies on <snip> the absence?
> As far as NFS consistency guarantees, they are described in the man nfs
> help page, as well as discussed in this thread
> <https://serverfault.com/questions/638601/nfs-caching-issue>.
> In short, without config changes that kill the performance, the NFS has:
>
>    - no cache coherency between clients for open file writes/reads. MPI
>    applications with concurrent I/O to the same file are affected by this, and
>    that leads to the development of parallel file systems like GPFS, Lustre,
>    PVFS. Toil is not affected by this because the client/server pipe on the
>    NFS always does full sync when the file is closed, and Toil does not do
>    concurrent writes (I hope).
>
>
Correct, the fileStore has always assumed immediate consistency, atomic
deletes and mvs, but definitely no attempts to do concurrent writing!


>
>    - no cache coherency for metadata operations (directory
>    modifications). Toil is affected by this, but not all is bad. NFS (see the
>    man page) provides strict time guarantees for how long it will take for the
>    cache to be reconciled on the client. These can be tuned in /etc/fstab.
>    Incidentally, the 30 sec max wait that I set in the spin-wait code is the
>    default in the NFS man page. Most sites never change the default.
>
> A number of workflow systems implement the spin-wait for this very pattern
> (Makeflow, Galaxy, Cromwell), and do just fine. Basically, the system
> learns that the job has finished when its disappears from qstat output,
> and the system waits for the output files to appear on the NFS. Consider
> also that the batch job scheduling system has its own delays in starting
> jobs, so the extra worst-case 30 sec (with default NFS setup) of wait will
> be incremental to the queue delays (probably double the delay).
>
> It is possible of course to implement some other notification mechanism
> like running a 0MQ server process in the leader, but is it really worth the
> trouble?
>

Agreed, almost certainly not.

> I agree that the spin-wait time should be made a tunable parameter.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/2068#discussion_r169528200>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkOINJ56MJkRVT16gHHBgnhDeTaPvfTks5tW4-RgaJpZM4SHi3T>
> .
>



-- 
Benedict
(calendar invites: bpaten@ucsc.edu,
 appointments: Tina Bernard <tibernar@ucsc.edu>)

diekhans commented:
Is the jobFile the only file that toil uses to communicate?  If so, this seems safe for Toil.  Of course, It will not prevent synchchonization with application's use of thes files.

diekhans commented:
We chatted with an admin with a lot of NFS experience and I think we view this as being useful.   @andreyto, would you be able to polish it up?  The rest of the group is pretty slammed with cloud deliverables right now.

diekhans commented on src/toil/jobStores/fileJobStore.py:
```diff
@@ -121,8 +122,18 @@ def batch(self):
 
-    def exists(self, jobStoreID):
-        return os.path.exists(self._getJobFileName(jobStoreID))
+    def exists(self, jobStoreID,maxTries=1):
```
I think it would be clearer to provide a separate function whose job it is to check for it a file that might appear vs just exists when called.  This caused some confusion with people looking at the code when this would happen.

Maybe waitForExists ?

andreyto commented:
@diekhans - yes, I will be able to polish this change as requested - add parameters for max wait time, factor it into a separate function, only print the warning on the first and last try to reduce chatter etc. I should have time to finish this in March.

However, what everyone said so far on this thread got me worried about using Toil on NFS even after this fix. Especially, what @hannes-ucsc and @benedictpaten said. 

It would be very useful if someone here who was the original designer would have created a small Wiki page and described how the flow of data works in Toil - what leader does, what worker does, can worker create new Toil "jobs" (quoted to distinguish them from qsub jobs), how the leader learns about the new jobs...

It is one thing when a workflow crashes. Much worse if it completes and generates faulty outputs. With all those extra housekeeping jobs that CWL layer creates in Toil, it is important to be certain that the job dependencies are not violated due to faulty assumptions about the FS synchronization.

Other workflow systems do not depend entirely on the NFS for the flow control. They typically work like this (e.g. Galaxy): the master tracks jobs through qstat (or DRMAA) polling. When qstat reports that the job is done (or is no longer present), the master spin-waits for the output files on the NFS. If it hits a time-out, the master marks the job as failed. Other systems do not poll qstat, but wrap the qsub-submitted command into a script that generates a flag file on the NFS upon job completion (Makeflow, Cromwell). The master then polls for this file to appear. The downside of this approach is if the job is killed with signal 9 (e.g. queue wallclock limit is exceeded), the flag file never appears. In both approaches, the consistency is ensured because there is a single node that runs the master.

Toil Torque module monitors the queue jobs, like that, so I assumed that the pattern was similar. But if Toil workers can generate new tasks through the jobStore, and monitoring of the qstat or a jobStore by a single master is not always involved in decision-making about job completion and order, then there might be a lot of places to be plugged. It might be more reliable, after all, to modify the jobStore code in order to guarantee synchronicity through some ephemeral network communication scheme.

diekhans commented:
Thanks @andreyto.

Hannes and myself look through your PR together yesterday we
believe this will work fine.  Part of the concern was confusion
on where the spin-wait applied.

Completely agree that Toil needs better design documents.  This
raises some points that should be included.

Yes. Toil uses the underlying batch system to monitor the job
state.  It uses the job store to communicate more information
back to the leader.  This information includes follow on jobs to
create.  Only the leader interacts with the batch scheduler, so
we are safe here.

Toils approach to no not assuming of a shared file system helps
with isolating this exchange.  Your approach will remove the
only known dependency of this timing of job completion and file
system in the core Toil framework.

Perhaps @mr-c can comment if there are any issues within the CWL
layer?







Andrey Tovchigrechko <notifications@github.com> writes:
> @diekhans - yes, I will be able to polish this change as requested - add
> parameters for max wait time, factor it into a separate function, only print
> the warning on the first and last try to reduce chatter etc. I should have time
> to finish this in March.
> 
> However, what everyone said so far on this thread got me worried about using
> Toil on NFS even after this fix. Especially, what @hannes-ucsc and
> @benedictpaten said.
> 
> It would be very useful if someone here who was the original designer would
> have created a small Wiki page and described how the flow of data works in Toil
> - what leader does, what worker does, can worker create new Toil "jobs" (quoted
> to distinguish them from qsub jobs), how the leader learns about the new
> jobs...
> 
> It is one thing when a workflow crashes. Much worse if it completes and
> generates faulty outputs. With all those extra housekeeping jobs that CWL layer
> creates in Toil, it is important to be certain that the job dependencies are
> not violated due to faulty assumptions about the FS synchronization.
> 
> Other workflow systems do not depend entirely on the NFS for the flow control.
> They typically work like this (e.g. Galaxy): the master tracks jobs through
> qstat (or DRMAA) polling. When qstat reports that the job is done (or is no
> longer present), the master spin-waits for the output files on the NFS. If it
> hits a time-out, the master marks the job as failed. Other systems do not poll
> qstat, but wrap the qsub-submitted command into a script that generates a flag
> file on the NFS upon job completion (Makeflow, Cromwell). The master then polls
> for this file to appear. The downside of this approach is if the job is killed
> with signal 9 (e.g. queue wallclock limit is exceeded), the flag file never
> appears. In both approaches, the consistency is ensured because there is a
> single node that runs the master.
> 
> Toil Torque module monitors the queue jobs, like that, so I assumed that the
> pattern was similar. But if Toil workers can generate new tasks through the
> jobStore, and monitoring of the qstat or a jobStore by a single master is not
> always involved in decision-making about job completion and order, then there
> might be a lot of places to be plugged. It might be more reliable, after all,
> to modify the jobStore code in order to guarantee synchronicity through some
> ephemeral network communication scheme.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

ejacox commented:
Thank you @diekhans and @hannes-ucsc for taking a look at this.

andreyto commented:
OK, thanks! I will work on finalizing it as soon as I can.

mr-c commented on src/toil/cwl/cwltoil.py:
```diff
@@ -742,8 +747,8 @@ def run(self, fileStore):
-                                    jobobj[key] = (shortname(inp["source"]),
-                                                   promises[inp["source"]].rv())
+                                    src = simplify_list(inp["source"])
+                                    jobobj[key] = (shortname(src), promises[src].rv())
```
https://github.com/BD2KGenomics/toil/pull/2069

mr-c commented on src/toil/cwl/cwltoil.py:
```diff
@@ -796,7 +801,8 @@ def run(self, fileStore):
         for out in self.cwlwf.tool["outputs"]:
-            outobj[shortname(out["id"])] = (shortname(out["outputSource"]), promises[out["outputSource"]].rv())
+            src = simplify_list(out["outputSource"])
+            outobj[shortname(out["id"])] = (shortname(src), promises[src].rv())
```
I think this was fixed by https://github.com/BD2KGenomics/toil/pull/2069 ; if so then I prefer that solution as it ensures all contributing links are processed correctly. If you have a complete example that doesn't work with that PR please share it so I can evaluate this in conjunction

DailyDreaming commented:
@andreyto What is the status on this?

andreyto commented:
@DailyDreaming I have rebased this and updated the spin-wait patch as was requested (created a separate function). I have added a couple of unrelated fixes that accumulated in the meantime. 

https://github.com/BD2KGenomics/toil/pull/2068/commits/8f7f568b0102d595ded6567630c7b7ec2ab68bcb addresses #2229. 

https://github.com/BD2KGenomics/toil/pull/2068/commits/c96e147043f3970e05d261db5fcf1a24c7b2c665 is still needed even on top of the latest code, but I could not isolate a small CWL snippet that causes it or figure out why it appears. I can share my entire workflow if it might help.

https://github.com/BD2KGenomics/toil/pull/2068/commits/f2e133991897cfbf12ca408cef557f083f1ce806 is a work-around to speed-up deletions of CWL temp output dirs after the workflow has completed. cwltool by default does it after each job, but enabling such immediate cleanup in cwltoil code causes exceptions related to missing outputs

mr-c commented:
@DailyDreaming is Jenkins down?

andreyto commented:
Jenkins seems to be stalled here? Can someone restart it before this pull request diverges from the master and has to be rebased again? Thanks!

DailyDreaming commented:
Jenkins, test this please.

DailyDreaming commented:
Jenkins, test this please.

benedictpaten commented:
What's up with this pr?

cgcloud-jenkins commented:
Can one of the admins verify this patch?

cgcloud-jenkins commented:
Build finished. 18 tests run, 0 skipped, 2 failed.

cgcloud-jenkins commented:
Build finished. 794 tests run, 33 skipped, 8 failed.

andreyto commented:
Folks, I cannot keep resolving conflicts that accumulate because the pull request is not getting merged for months. And the failed test results are not accessible.

DailyDreaming commented:
@andreyto I've resolved the conflict.  Sorry this hasn't been looked at in a while.

Currently tests are posted at:
http://jenkins.cgcloud.info/job/toil-pull-requests/
http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/
This is temporary for the moment since the DataBiosphere move.

Once tests are run, I'll post links to the results.  I can try to resolve any failing tests if you like.

cgcloud-jenkins commented:
Build finished. 882 tests run, 33 skipped, 137 failed.

DailyDreaming commented:
The 137 failed tests: http://jenkins.cgcloud.info/job/toil-pull-requests/4304/

cgcloud-jenkins commented:
Build finished. 880 tests run, 33 skipped, 1 failed.

DailyDreaming commented:
One failing test: http://jenkins.cgcloud.info/job/toil-pull-requests/4306/testReport/junit/src.toil.test.src.resourceTest/ResourceTest/testBuiltIn/

Will look at this tomorrow.

DailyDreaming commented:
Jenkins, test this please.

cgcloud-jenkins commented:
Build finished. 18 tests run, 0 skipped, 2 failed.

cgcloud-jenkins commented:
Build finished. 880 tests run, 33 skipped, 1 failed.

DailyDreaming commented:
Same test: http://jenkins.cgcloud.info/job/toil-pull-requests/4309/testReport/junit/src.toil.test.src.resourceTest/ResourceTest/testBuiltIn/

@natanlao ran this locally and it seems to pass.  >.<

natanlao commented:
Update: `ResourceTest.testBuiltIn` will pass if run individually, but will fail if run via `make test_parallel`, with a similar traceback:

```
__________________________________ ResourceTest.testBuiltIn ___________________________________
Traceback (most recent call last):
  File "/home/natan/cgl/andreyto-toil/src/toil/test/src/resourceTest.py", line 107, in testBuiltIn
    self._test(module_name, shouldBelongToToil=True)
  File "/home/natan/cgl/andreyto-toil/src/toil/test/src/resourceTest.py", line 128, in _test
    resource = module.saveAsResourceTo(jobStore)
  File "/home/natan/cgl/andreyto-toil/src/toil/resource.py", line 457, in saveAsResourceTo
    return self._getResourceClass().create(jobStore, self._resourcePath)
  File "/home/natan/cgl/andreyto-toil/src/toil/resource.py", line 87, in create
    with cls._load(leaderPath) as src:
  File "/home/natan/cgl/andreyto-toil/src/toil/resource.py", line 294, in _load
    zipFile.write(fullPath, os.path.relpath(fullPath, rootDir))
  File "/usr/lib/python2.7/zipfile.py", line 1123, in write
    st = os.stat(filename)
OSError: [Errno 2] No such file or directory: '/home/natan/cgl/andreyto-toil/src/toil/test/cwl/spec/v1.0/tmp/1/b/out_tmpdirXTNyDU/ref.fasta
```
This leads me to believe there is some kind of test isolation issue.

natanlao commented:
I can reproduce the test failure on a fresh clone with `make test_parallel tests="src/toil/test/cwl/cwlTest.py src/toil/test/src/resourceTest.py::ResourceTest::testBuiltIn"`

cgcloud-jenkins commented:
Build finished. 880 tests run, 39 skipped, 1 failed.

cgcloud-jenkins commented:
Build finished. 880 tests run, 39 skipped, 0 failed.

andreyto commented:
Wonderful, thanks a lot, everyone!
